### ISSUE TITLE
Nested Update/Effects

### ISSUE THREAD

**Issue by dodekeract** [hyperapp/hyperapp#73]
Have you considered this:

```javascript
app({
	update: {
		example: {
			attempt: () => alert(1)
		}
	}
})
```

It should then be possible to call `msg.example.attempt()`.

---

<details>
<summary>Meanwhile I ended up using this</summary>

```javascript
const convert = (actions, prefix = '') => Object
	.keys(actions)
	.map(name => typeof actions[name] === 'function'
		? ({[`${prefix[0].toLowerCase()}${prefix.slice(1)}${name[0].toUpperCase()}${name.slice(1)}`]: actions[name]})
		: (convert(actions[name], `${prefix}${name[0].toUpperCase()}${name.slice(1)}`))
	)
	.reduce((a, b) => Object.assign(a, b), {})

app({
	update: convert({
		example: {
			attempt: () => alert(1)
		}
	})
})
```

This maps `example.attempt` to `exampleAttempt`.

</details>

**Comment by jbucaran** [comment]
@dodekeract Interesting. Nope. I never considered this.

Sometimes because it's nice to have doesn't mean we should have it, but I am open to feedback.

**Comment by tunnckoCore** [comment]
Don't see value. It is better to be flat. But any use case, more real examples?

**Comment by dodekeract** [comment]
@tunnckoCore basically anything with lots of updates/effects. That way it'd be possible to split it into multiple files.

**Comment by jbucaran** [comment]
@dodekeract 

```js
const red = require("./red-effects")
const blue = require("./blue-effects")
const yellow = require("./yellow-effects")

app({
	update: Object.assign({}, red, blue, yellow), ...
})
```

Does that make sense_?

**Comment by tunnckoCore** [comment]
@jbucaran it makes, but there may be name collisions... That's the important thing.

**Comment by jbucaran** [comment]
choo solved this with namespaces IIRC.

I didn't like the idea of namespaces, so I didn't implement any. 

What other options do we have?

**Comment by tunnckoCore** [comment]
> choo solved this with namespaces IIRC.

Yes and no. Not exactly, or at least I think that they have the same problem in that case. There namespaces are for different models.

**Comment by dodekeract** [comment]
@tunnckoCore We can talk about combining this with the solution for multiple models/components/etc. though.

**Comment by zaceno** [comment]
Vuex (Vue.js equivalent to `msg`) also uses a flat structure. The reasoning is, as far as I understand it, that there is no practical benefit to namespacing/nesting over simply prefixing as a strategy to avoid naming collisions.

That said, I also am curious about strategies (in general -- not just for the reducers/effects) for structuring and sensibly managing larger scale, complex apps (with potentially hundreds of reducers). All my experience with really big and complex apps is from the world of MVC, and in coming to this brave new world of the Elm-ish/React-ish paradigm, I have mostly been exposed to trivial examples and haven't found much in terms of composing larger apps from smaller pieces (like a SAM equivalent of H(ierarchical)MVC for example).

**Comment by tunnckoCore** [comment]
@zaceno same here :) I was MVC a long ago, but still can get the Components thing really enough. But also don't have enough time and wish to review some big project using vue, react or whatever.

**Comment by selfup** [comment]
```
const red = require("./red-effects")
const blue = require("./blue-effects")
const yellow = require("./yellow-effects")

app({
  update: { red, blue, yellow }
})
```

Now there is no naming conflict but it is no longer flat 🤔

**Comment by dodekeract** [comment]
@selfup This is possible though `update: {...red, ...blue, ...yellow}`.

**Comment by selfup** [comment]
@dodekeract Ohh that is nice too

**Comment by dodekeract** [comment]
@selfup the last one. In this case the one from `yellow`.

**Comment by selfup** [comment]
Oh interesting! Good to know there is a fallback

**Comment by jbucaran** [comment]
Closing as I don't think I'll be implementing nested actions.

**Comment by nichoth** [comment]
I ended up using the same nested interface in my project also to compose components, seems pretty good so far.

**Comment by jbucaran** [comment]
Forgot to close here. This has been implemented by @dodekeract and it's available in 0.6.0 🎉!

/cc @nichoth
